### PR TITLE
feat: add mod update checker to Registry system

### PR DESCRIPTION
**Pull request by Enduriel**
Pretty simple feature in theory, modders writes
```squirrel
::MSU.Mod.Registry.setGithubURL("https://github.com/MSUTeam/MSU");
```
and optionally also
```squirrel
::MSU.Mod.Registry.setNexusModsURL("https://www.nexusmods.com/battlebrothers/mods/479");
```
These are verified as correct URLs for the sites using regexes defined in registry_mod_addon.

Mod gets GitHub (and maybe one day Nexusmods as well if they either make their API public or we store player keys with cookies) tags for all mods with a defined URL and checks them against local versions (if the online versions aren't using semver it lets the player know which mod is failing the check) and if there are updates available those get pushed to the player in a neat little addon to Popup.
image
Dummy in this example doesn't have a nexusmods URL set so the button is grayscale.

The one thing I'm still not 100% on is whether the buttons aren't obvious enough as download links, so I'd appreciate feedback there.

### CODE REVIEW

File: msu/systems/registry/registry_mod_addon.nut [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	{
+		return this.__GithubURL;
+	}
+
+	function setGithubURL( _url )
+	{
+		::MSU.requireString(_url);
+		if (!this.GithubURLRegex.match(_url))
+		{
+			::logError("A Github link must be a link to a specific repository, eg: https://github.com/MSUTeam/MSU");
```
Comment (TaroEld) [inline msu/systems/registry/registry_mod_addon.nut:19]: Probably add something like "Is there an issue with your URL"?

File: msu/systems/registry/registry_system.nut [hunk 2]
```diff
@@ -37,6 +41,47 @@
...
+			if (!::MSU.SemVer.isSemVer(version))
+			{
+				::MSU.Popup.showRawText(format("The version '%s' from the mod %s (%s) isn't using <span style=\"color: lightblue; text-decoration: underline;\" onclick=\"openURL('https://semver.org')\">semantic versioning</span> for its online versions, despite registering to use the MSU update checker. Let the mod author know if you see this error", version,  mod.getName(), modID));
+				continue;
+			}
+			// TODO check if mod version is regex, or if the mod author has fucked something up (THIS IS IMPORTANT)
+			if (!::MSU.SemVer.compareVersionWithOperator(version, ">", mod)) continue;
+			local type = "PATCH";
+			if (::MSU.SemVer.compareMajorVersionWithOperator(version, ">", mod)) type = "MAJOR";
+			else if (::MSU.SemVer.compareMinorVersionWithOperator(version, ">", mod)) type = "MINOR";
```
Comment (TaroEld) [inline msu/systems/registry/registry_system.nut:72]: Enum those?

File: scripts/config/msu/string.nut [hunk 3]
```diff
@@ -27,4 +27,14 @@
...
+
+	function startsWith( _string, _start )
+	{
+		return _string.find(_start) == 0;
+	}
+
+	function endsWith( _string, _end )
+	{
+		return _end.len() <= _string.len() && _string.slice(-_end.len()) == _end;
+	}
```
Comment (TaroEld) [inline scripts/config/msu/string.nut:39]: Let me be the stickler this time and say that this belongs in a different commit

File: ui/mods/msu/popup.js [hunk 4]
```diff
@@ -96,25 +100,62 @@ MSUPopup.prototype.showRawText = function (_data)
...
+		var modInfoContainer = $('<div class="msu-mod-info-container"/>');
+		self.mListScrollContainer.append(modInfoContainer)
+		modInfoContainer.append($('<div class="mod-name title title-font-big font-bold font-color-title">' + _modInfo.name + '</div>'));
+		var githubContainer = $('<div class="l-github-button"/>');
+		modInfoContainer.append(githubContainer);
+		var githubButton = githubContainer.createImageButton(Path.GFX + "mods/msu/logos/github-32.png", function ()
+		{
+			openURL(_modInfo.githubURL);
+		});
+		if (_modInfo.githubURL == null) githubButton.attr('disabled', true);
```
Comment (TaroEld) [inline ui/mods/msu/popup.js:132]: It's black and white so this is probably not very visible. Just hide them if they're inactive?

**Review by TaroEld — CHANGES_REQUESTED** [review]
Cool stuff, some small changes

File: scripts/config/msu/string.nut (hunk 3 above)
Comment (Enduriel) [reply scripts/config/msu/string.nut:39]: That's fair, not sure how this happened.

File: ui/mods/msu/popup.js (hunk 4 above)
Comment (Enduriel) [reply ui/mods/msu/popup.js:132]: That's fair ig, though it'd never come up rn

Comment (TaroEld) [reply ui/mods/msu/popup.js:132]: Yeah ik but for completion's sake

File: msu/systems/registry/registry_system.nut (hunk 2 above)
Comment (Enduriel) [reply msu/systems/registry/registry_system.nut:72]: I'm not sure, these don't come up very often and I'd need to enum them again JS side as well which is ugly I think?

File: msu/systems/registry/registry_mod_addon.nut (hunk 1 above)
Comment (Enduriel) [reply msu/systems/registry/registry_mod_addon.nut:19]: Sure

File: msu/systems/registry/registry_system.nut (hunk 2 above)
Comment (TaroEld) [reply msu/systems/registry/registry_system.nut:72]: Yeah sure, not important.

File: msu/msu_mod/setup_msu_mod.nut [hunk 5]
```diff
@@ -1 +1,3 @@
 ::MSU.Mod <- ::MSU.Class.Mod(::MSU.ID, ::MSU.Version, ::MSU.Name);
```
Comment (LordMidas) [inline msu/msu_mod/setup_msu_mod.nut:1]: These should not be part of this PR because this PR adds the framework. Having it in the PR is probably not a big deal but the way you've got it set up in the commits right now is that this addition happens in the same commit that adds this framework. The addition of MSU links should be a separate commit.

File: msu/systems/registry/registry_mod_addon.nut [hunk 6]
```diff
@@ -0,0 +1,63 @@
+::MSU.Class.RegistryModAddon <- class extends ::MSU.Class.SystemModAddon
+{
+	static GithubURLRegex = regexp("https:\\/\\/github\\.com\\/([-\\w]+)\\/([-\\w]+)");
+	static NexusURLRegex = regexp("https:\\/\\/www\\.nexusmods\\.com\\/battlebrothers\\/mods\\/(\\d+)");
+	__GithubURL = null;
+	__NexusModsURL = null;
+	__UpdateSource = null;
+
+	function getGithubURL()
```
Comment (LordMidas) [inline msu/systems/registry/registry_mod_addon.nut:9]: The official spelling of GitHub is with `H` being capital. Please fix throughout the PR.

File: msu/systems/registry/registry_system.nut [hunk 7]
```diff
@@ -37,6 +41,47 @@
...
+		local modsWithNewVersions = {};
+		foreach (modID, version in _modVersions)
+		{
+			local mod = ::MSU.System.Registry.getMod(modID);
+			if (!::MSU.SemVer.isSemVer(version))
+			{
+				::MSU.Popup.showRawText(format("The version '%s' from the mod %s (%s) isn't using <span style=\"color: lightblue; text-decoration: underline;\" onclick=\"openURL('https://semver.org')\">semantic versioning</span> for its online versions, despite registering to use the MSU update checker. Let the mod author know if you see this error", version,  mod.getName(), modID));
+				continue;
+			}
+			// TODO check if mod version is regex, or if the mod author has fucked something up (THIS IS IMPORTANT)
```
Comment (LordMidas) [inline msu/systems/registry/registry_system.nut:68]: I'd like to politely request to abstain from such language in MSU code.

File: msu/systems/registry/registry_mod_addon.nut [hunk 8]
```diff
@@ -0,0 +1,63 @@
+::MSU.Class.RegistryModAddon <- class extends ::MSU.Class.SystemModAddon
```
Comment (LordMidas) [inline msu/systems/registry/registry_mod_addon.nut:1]: Overall, I am wondering if it is better to have a single `addURL` function exposed to the public API instead of separate functions for different websites e.g. `addGitHubURL` and `addNexusModsURL`. We may want to have another website in the future, and this adds functions for each website. With a single `addURL` function we can use the regex to automatically detect what domain the URL belongs to, and if it is github we set it as `__GitHubURL` and and respectively for other internal variables.

**Review by LordMidas — CHANGES_REQUESTED** [review]

File: msu/systems/registry/registry_mod_addon.nut (hunk 8 above)
Comment (Enduriel) [reply msu/systems/registry/registry_mod_addon.nut:1]: I don't really wanna do that since it makes it harder to give a useful error msg, if we make a LinkSource enum or smtng and then pass those together that would be fine by me but idk if that's really much better...

File: msu/systems/registry/registry_system.nut (hunk 7 above)
Comment (Enduriel) [reply msu/systems/registry/registry_system.nut:68]: That was meant to be temporary, my bad

**Review by TaroEld — APPROVED** [review]

File: msu/msu_mod/msu_registry.nut [hunk 9]
```diff
@@ -0,0 +1,3 @@
+::MSU.Mod.Registry.addModSource(::MSU.System.Registry.URLDomain.GitHub, "https://github.com/MSUTeam/MSU");
```
Comment (LordMidas) [inline msu/msu_mod/msu_registry.nut:1]: As previously requested, please move this (adding MSU links) out to a separate PR as this PR should be solely the framework itself.

File: msu/systems/registry/load.nut [hunk 10]
```diff
@@ -2,12 +2,18 @@ local function includeFile( _file )
...
-includeFile("registry_system.nut");
+includeFile("registry_system");
+includeFile("registry_mod_addon");
 
 local system = ::MSU.Class.RegistrySystem();
-
 ::MSU.System.Registry <- system;
 
+includeFile("mod_source");
+includeFile("mod_source_github");
```
Comment (LordMidas) [inline msu/systems/registry/load.nut:12]: Not a big deal but probably better to have a `mod_sources` subfolder. But I don't mind if we leave it like this for now.

File: msu/systems/registry/mod_source_github.nut [hunk 11]
```diff
@@ -0,0 +1,21 @@
+::MSU.Class.ModSourceGitHub <- class extends ::MSU.Class.ModSource
+{
+	static URLDomain = ::MSU.System.Registry.URLDomain.GitHub;
+	static Regex = regexp("https:\\/\\/github\\.com\\/([-\\w]+)\\/([-\\w]+)");
+
+	constructor( _url )
+	{
+		if (!this.Regex.match(_url))
+		{
+			::logError("A GitHub link must be a link to a specific repository, eg: 'https://github.com/MSUTeam/MSU' Check to make sure there's not an issue with your URL and that it is formatted the same way as the MSU URL.");
```
Comment (LordMidas) [inline msu/systems/registry/mod_source_github.nut:10]: The official spelling of `eg` is `e.g.`

File: msu/systems/registry/mod_source_nexusmods.nut [hunk 12]
```diff
@@ -0,0 +1,20 @@
+::MSU.Class.ModSourceNexusMods <- class extends ::MSU.Class.ModSource
+{
+	static URLDomain = ::MSU.System.Registry.URLDomain.NexusMods;
+	static Regex = regexp("https:\\/\\/www\\.nexusmods\\.com\\/battlebrothers\\/mods\\/(\\d+)");
+
+	constructor( _url )
+	{
+		if (!this.Regex.match(_url))
+		{
+			::logError("A NexusMods link must be a link to a specific mod's main page, eg: 'https://www.nexusmods.com/battlebrothers/mods/479' Check to make sure there's not an issue with your URL and that it is formatted the same way as the MSU URL.");
```
Comment (LordMidas) [inline msu/systems/registry/mod_source_nexusmods.nut:10]: The official spelling of `eg` is `e.g.`

File: msu/systems/registry/registry_mod_addon.nut [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+	function hasModSource( _domain )
+	{
+		return _domain in this.__ModSources;
+	}
+
+	function addModSource( _domain, _url )
+	{
+		::MSU.requireString(_url);
+		this.__ModSources[_domain] <- ::MSU.System.Registry.ModSources[_domain](_url)
```
Comment (LordMidas) [inline msu/systems/registry/registry_mod_addon.nut:19]: Missing semi-colon

File: msu/systems/registry/registry_system.nut [hunk 14]
```diff
@@ -1,15 +1,27 @@
...
 	Mods = null;
 	constructor()
 	{
 		base.constructor(::MSU.SystemID.Registry);
 		this.Mods = {};
 	}
 
+	function addNewModSource( _modSource )
+	{
+		::MSU.requireInstanceOf(::MSU.Class.ModSource, _modSource.instance());
```
Comment (LordMidas) [inline msu/systems/registry/registry_system.nut:17]: Why `_modSource.instance()` and not `_modSource` only?

**Review by LordMidas — CHANGES_REQUESTED** [review]
Awesome work @Enduriel! Really nice. Overall big fan of this PR and the way it is organized now. I've marked the few comments that I have in my review.

File: msu/systems/registry/registry_system.nut (hunk 14 above)
Comment (Enduriel) [reply msu/systems/registry/registry_system.nut:17]: Because you can't `instanceof` against a class, only an instance, `.instance()` generates a dummy instance without calling the constructor

Comment (LordMidas) [reply msu/systems/registry/registry_system.nut:17]: No because the `::MSU.requireInstanceOf` already does a `typeof instance` check on the param. So you don't need to do `.instance()`.

File: msu/msu_mod/msu_registry.nut (hunk 9 above)
Comment (LordMidas) [reply msu/msu_mod/msu_registry.nut]: I've fixed this part.

File: msu/systems/registry/registry_system.nut [hunk 15]
```diff
@@ -1,15 +1,27 @@
 ::MSU.Class.RegistrySystem <- class extends ::MSU.Class.System
 {
+	static ModSources = {};
+	static URLDomain = ::MSU.Class.Enum([
```
[the remark this answers is not in the document]
Comment (Enduriel) [reply msu/systems/registry/registry_system.nut:4]: Yeah, that's fair

**Review by LordMidas — APPROVED** [review]
